### PR TITLE
Usable logic tests (works now!)

### PR DESCRIPTION
**Pull request by mariusandra**
## Where are the logic tests?

I've been punting on logic tests for a while because something just hasn't clicked with them. The way [currently built](https://github.com/PostHog/posthog/pull/5234/files#diff-56eb9535637e0a30f7d65f04a1c124b3c16f92966904bbef27ab28aa9a654fd1) (or in an adjacent [random experiment](https://github.com/PostHog/posthog/pull/5846)), they would become a development crutch (like E2E tests), and not a useful aid... and I couldn't find a way to bring all of this together. I don't want to push for an approach I don't agree with, hence silence.

That changes now. This PR proposes a simple syntax **and implementation** for logic tests, with a mental model that in my opinion makes building them obviously clear.

## Mental model

The mental model here is to think of a **logic == a collection stories, but mixed up**. 

Take dashboards. You can do a lot of things with them: rename them, duplicate them, etc. These are all stories. While developing, your "make it so we can duplicate dashboards" story is clear, but over time as more and more stories get added to the mix, the dashboard logic becomes too nosiy to follow.

Screenshot 2021-09-16 at 12 46 42

There is a lot of value in having the "when reading the code after" final form of a logic, with all actions, selectors, etc grouped together. However the original stories, the original intent, is lost.

Hence --> each time you add something, document the story in a test and make your changes to the logic

Screenshot 2021-09-16 at 12 53 45

It's the same story, but written twice, one side checking the other. Two factor authentication for your code.

## Code

Luckily for us, a history of Redux actions reads like a story. What follows is actually runnable test code from this PR, optimised for pattern matching. Notice the interplay between code with `await` and without.

- The approach can be summarised as a "redux action record/playback" implementation. 
- We record all actions and state changes, and let you conveniently query the recorded flow
- If an action hasn't been dispatched yet, add an `await` and the same code will magically wait for it before checking the next values, otherwise everything runs sync
- You can check the state of values between actions, even if the current values have already changes

```ts
describe('infiniteListLogic verbose version', () => {
    let logic: BuiltLogic<infiniteListLogicType>

    mockAPIGet(async ({ pathname, searchParams }) => {
        // skipped for brevity
    })

    initKeaTestLogic({
        logic: infiniteListLogic,
        props: {
            taxonomicFilterLogicKey: 'testList',
            listGroupType: TaxonomicFilterGroupType.Events,
        },
        onLogic: (l) => (logic = l), // todo: use a proxy and get rid of this line
    })

    it('calls loadRemoteItems on mount', async () => {
        await expectLogic(logic)
            .toDispatchActions(['loadRemoteItems', 'loadRemoteItemsSuccess'])
            .toMatchValues({
                remoteItems: expect.objectContaining({
                    results: expect.arrayContaining([expect.objectContaining({ name: 'event1' })]),
                }),
            })
    })

    it('setting search query loads remote items', async () => {
        await expectLogic(logic, () => logic.actions.setSearchQuery('event'))
            .toDispatchActions(['setSearchQuery', 'loadRemoteItems'])
            .toMatchValues({
                searchQuery: 'event',
                remoteItems: expect.objectContaining({
                    count: 56, // old values from initial load
                }),
                remoteItemsLoading: true,
            })
            .toDispatchActions(['loadRemoteItemsSuccess'])
            .toMatchValues({
                searchQuery: 'event',
                remoteItems: expect.objectContaining({
                    count: 3, // got new results
                    results: expect.arrayContaining([expect.objectContaining({ name: 'event1' })]),
                }),
                remoteItemsLoading: false,
            })
    })

    it('setting search query filters events', async () => {
        await expectLogic(logic, () => logic.actions.setSearchQuery('event'))
            .toDispatchActions(['setSearchQuery', 'loadRemoteItems', 'loadRemoteItemsSuccess'])
            .toMatchValues({
                searchQuery: 'event',
                remoteItems: expect.objectContaining({
                    count: 3,
                    results: expect.arrayContaining([expect.objectContaining({ name: 'event1' })]),
                }),
            })
    })

    describe('index', () => {
        it('is set via setIndex', async () => {
            expectLogic(logic).toMatchValues({ index: 0 })
            expectLogic(logic, () => logic.actions.setIndex(1)).toMatchValues({ index: 1 })
        })

        it('can go up and down', async () => {
            await expectLogic(logic).toDispatchActions(['loadRemoteItems', 'loadRemoteItemsSuccess'])
            expectLogic(logic).toMatchValues({ index: 0, remoteItems: expect.objectContaining({ count: 56 }) })
            expectLogic(logic, () => logic.actions.moveUp()).toMatchValues({ index: 55 })
            expectLogic(logic, () => logic.actions.moveUp()).toMatchValues({ index: 54 })
            expectLogic(logic, () => logic.actions.moveDown()).toMatchValues({ index: 55 })
            expectLogic(logic, () => logic.actions.moveDown()).toMatchValues({ index: 0 })
            expectLogic(logic, () => logic.actions.moveDown()).toMatchValues({ index: 1 })
            expectLogic(logic, () => logic.actions.moveUp()).toMatchValues({ index: 0 })
        })
    })

    it('selects the selected item', async () => {
        expectLogic(logic).toMatchValues({ selectedItem: expect.objectContaining({ name: 'event1' }) })
        expectLogic(logic, () => logic.actions.selectSelected()).toDispatchActions([
            logic.actionCreators.selectSelected(),
            logic.actionCreators.selectItem(
                TaxonomicFilterGroupType.Events,
                'event1',
                expect.objectContaining({ name: 'event1' })
            ),
        ])
    })
})

```

image

## The API

Run `expectLogic(logic: BuiltLogic, callback: (logic) => void | Promise<void>)` with an optional `callback` and append [any of the following functions](https://github.com/PostHog/posthog/blob/a58191764d4291782ee3a9905d6f6442a58eb344/frontend/src/test/kea-test-utils/functions/index.ts#L11):

```ts
export interface ExpectLogicMethods {
    then: (callback?: (value: any) => void | Promise<void>) => Promise<void>
    toDispatchActions: (actions: ActionToDispatch[]) => ExpectLogicMethods
    toFinishListeners: (arg?: undefined) => ExpectLogicMethods
    toFinishAllListeners: (arg?: undefined) => ExpectLogicMethods
    toMatchValues: (values: Record<string, any>) => ExpectLogicMethods
    toMount: (otherLogics?: BuiltLogic | LogicWrapper | (BuiltLogic | LogicWrapper)[]) => ExpectLogicMethods
    printActions: (payload?: string) => ExpectLogicMethods
    delay: (ms: number) => ExpectLogicMethods
}
```

- [x] Extend the kea-waitfor plugin into a test aide, capture all events as they're dispatched (the plugin discards the past and only awaits from this moment, in the tests we need to check what ran a few cycles ago)
- [x] Build the `expectLogic` function to await for the `.toDispatchActions` actions array in order
- [x] Build the `.toDispatchActions` and `.toMatchValues` functions
- [x] Build "`await` for listeners to finish" and add it to `.toMatchValues` if there's no match
- [x] Make all of this work seamlessly with async code
- [x] Clean up the code behind `expectLogic`
- [x] Better examples with multiple logics changing data on themselves
- [ ] Get a more robust object comparison system than `objectsEqual`

## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests
- [ ] Migrations are safe to run at scale (e.g. PostHog Cloud) – present proof if not obvious
- [ ] New/changed UI is decent on smartphones (viewport width around 360px)

### CODE REVIEW

**Comment by mariusandra** [comment]
Hey all, this is ready for an initial look. Don't be intimidated by the number of changed files. The important example test files to look at are [infiniteListLogic.test.ts](https://github.com/PostHog/posthog/blob/90dc09ad01644bb2759cecbea51f4c98fe95daf7/frontend/src/lib/components/TaxonomicFilter/infiniteListLogic.test.ts) and [funnelLogic.test.ts](https://github.com/PostHog/posthog/blob/90dc09ad01644bb2759cecbea51f4c98fe95daf7/frontend/src/scenes/funnels/funnelLogic.test.ts). 

When reviewing, see if everything there looks fine and the tests makes sense. Possibly also try to write a test as well, including try the other test functions like `printActions()`. There's a lot of other code around the tests to make `expectLogic` work, which should get secondary priority. It will eventually be extracted to a separate `kea-test-utils` package anyway.

The API mocking story here is still a work in progress, and I would appreciate any kind of approaches on how to improve it. However that's outside the scope of this PR.

The caveat with `funnelLogic` now is that it boots up a lot of other logics (e.g. preflight, insight history, etc) and makes testing hard because of many side effects and unmounted logics. There's thus stuff to clean up with the logics themselves.

**Comment by alexkim205** [comment]
Quick q — is there a utility function like that can print out all the logics that are mounted?

**Comment by mariusandra** [comment]
Not yet, though feel free to add one. Use `getContext()` in kea and then [print out `mounted` from this part](https://github.com/keajs/kea/blob/master/src/types.ts#L388-L391).

File: frontend/src/lib/utils.tsx [hunk 1]
```diff
@@ -53,7 +53,13 @@ export const ANTD_TOOLTIP_PLACEMENTS: Record<any, AlignType> = {
 
 export function uuid(): string {
     return '10000000-1000-4000-8000-100000000000'.replace(/[018]/g, (c) =>
-        (parseInt(c) ^ (crypto.getRandomValues(new Uint8Array(1))[0] & (15 >> (parseInt(c) / 4)))).toString(16)
+        (
+            parseInt(c) ^
+            ((typeof window?.crypto !== 'undefined'
+                ? window.crypto.getRandomValues(new Uint8Array(1))[0]
+                : Math.floor(Math.random() * 256)) &
```
Comment (Twixes) [inline frontend/src/lib/utils.tsx:60]: Would appreciate a comment for future context (browser v. Node)

File: frontend/src/scenes/funnels/funnelLogic.test.ts [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+            }
+        } else if (pathname.startsWith('api/insight')) {
+            return { results: [], next: null }
+        } else if (pathname === '_preflight/') {
+            return { is_clickhouse_enabled: true }
+        } else if (pathname === 'api/users/@me/') {
+            return {}
+        } else {
+            console.error(`Unmocked fetch to: ${pathname}`)
+            debugger
```
Comment (Twixes) [inline frontend/src/scenes/funnels/funnelLogic.test.ts:33]: What does `debugger` do in here?

File: frontend/src/scenes/insights/insightLogic.tsx [hunk 3]
```diff
@@ -241,7 +245,7 @@ export const insightLogic = kea<insightLogicType>({
                 const duration = new Date().getTime() - values.queryStartTimes[queryId]
                 const tags = {
                     insight: values.activeView,
-                    scene: sceneLogic.values.scene,
+                    scene: IS_TEST_MODE ? 'circular load if sceneLogic no loaded first' : sceneLogic.values.scene,
```
Comment (Twixes) [inline frontend/src/scenes/insights/insightLogic.tsx:248]: This doesn't look great… Is there no other way?

File: frontend/src/scenes/funnels/funnelLogic.test.ts (hunk 2 above)
Comment (mariusandra) [reply frontend/src/scenes/funnels/funnelLogic.test.ts]: Why it drops you into a debugger of course! Removed.

File: frontend/src/scenes/insights/insightLogic.tsx (hunk 3 above)
Comment (mariusandra) [reply frontend/src/scenes/insights/insightLogic.tsx]: Well, removed, but this does boot up `sceneLogic` and as a result loads everything that the insightsScene needs, which is a lot more extra stuff that suddenly appears. I guess adding a `.isMounted()` to the logic will solve the issue.

Until then, tests are just a bit slower:

image

**Review by mariusandra — COMMENTED** [review]
Feedback addressed. I also added a new mode, which lets you write multi-logic tests with a common action pointer (used to be per-logic)

```ts
it('has clickhouse enabled once preflight loads', async () => {
    await expectLogic() // no (logic) here, add one to use it as a default that gets overridden if needed
        .toDispatchActions(preflightLogic, ['loadPreflight'])
        .toMatchValues(logic, {
            clickhouseFeaturesEnabled: false,
        })
        .toDispatchActions(preflightLogic, ['loadPreflightSuccess'])
        .toMatchValues(logic, {
            clickhouseFeaturesEnabled: true,
        })
})
```

**Comment by mariusandra** [comment]
If nobody minds, I'll merge this when tests pass to:
1. Have a point we can iterate on
2. Already start writing logic tests for new functionality

**Pull request merged**